tombstone_gc.cc

```diff
@@ -78,32 +78,14 @@ void tombstone_gc_state::drop_repair_history_for_table(const table_id& id) {
 // 1) if the tombstone_gc_mode is not repair, since we have the same value for all the keys in the ranges.
 // 2) if the tombstone_gc_mode is repair, and the range is a sub range of a range in the repair history map.
 tombstone_gc_state::get_gc_before_for_range_result tombstone_gc_state::get_gc_before_for_range(schema_ptr s, const dht::token_range& range, const gc_clock::time_point& query_time) const {
-    bool knows_entire_range = true;
-    const auto& options = s->tombstone_gc_options();
-    switch (options.mode()) {
-    case tombstone_gc_mode::timeout: {
-        dblog.trace("Get gc_before for ks={}, table={}, range={}, mode=timeout", s->ks_name(), s->cf_name(), range);
-        auto gc_before = check_min(s, saturating_subtract(query_time, s->gc_grace_seconds()));
-        return {gc_before, gc_before, knows_entire_range};
-    }
-    case tombstone_gc_mode::disabled: {
-        dblog.trace("Get gc_before for ks={}, table={}, range={}, mode=disabled", s->ks_name(), s->cf_name(), range);
-        return {gc_clock::time_point::min(), gc_clock::time_point::min(), knows_entire_range};
-    }
-    case tombstone_gc_mode::immediate: {
-        dblog.trace("Get gc_before for ks={}, table={}, range={}, mode=immediate", s->ks_name(), s->cf_name(), range);
-        auto t = check_min(s, query_time);
-        return {t, t, knows_entire_range};
-    }
-    case tombstone_gc_mode::repair: {
-        const std::chrono::seconds& propagation_delay = options.propagation_delay_in_seconds();
+
+    auto gc_before_from_reconcile_stats = [this, &range, &s](seastar::lw_shared_ptr<reconcile_history_map> m, const std::chrono::seconds& propagation_delay) {
```

Comment:
What is all this logic calculating? Is it necessary for group 0 tables? I don't understand why we need anything more complex than comparing two integers.

In fact this super complex data structure:
```cpp
using reconcile_history_map = boost::icl::interval_map<dht::token, gc_clock::time_point, boost::icl::partial_absorber, std::less, boost::icl::inplace_max>;
```

seems completely unnecessary for group 0 tables. Unlike repair mode, we do not have to consider different token ranges separately. We have a single timestamp (extracted from state_id) that covers the full range of all keys. The code you wrote:
```cpp
void tombstone_gc_state::update_group0_refresh_time(gc_clock::time_point refresh_time) {
    auto m = get_or_create_group0_history();
    SCYLLA_ASSERT(m);
    *m += std::make_pair(boost::icl::interval<dht::token>::closed(dht::minimum_token(), dht::maximum_token()), refresh_time);
}```

is evidence. If you always update the full interval, why use interval_map in the first place.

Reply:
Yeah, I wanted to reuse the calculation as done for the "repair" mode. But in fact if we don't have to consider the ranges we can probably simplify this so it would make sense to keep these separately.

Reply:
reworked to keep the previous repair code under the repair handling and adding a separate simplified calculation for the group0